src/tools/rest-backend/benchmarks/benchmark_kdbrest_elektra.cpp

```diff
@@ -0,0 +1,462 @@
+{
+	std::cout << "Benchmark: Lookup single entry by fully specified key\n"
+		  << "          (" << numUsers << " users à " << numEntriesPerUser << " entries with " << numTagsPerEntry << " tags each)"
+		  << std::endl;
+	std::cout << "==============================================================" << std::endl;
+
+	std::cout << "-> Refreshing database (clear)" << std::endl;
+
+	clearDatabase ();
+
+	std::cout << "-> Creating test data (" << (numUsers + numUsers * numEntriesPerUser + numUsers * numEntriesPerUser * numTagsPerEntry)
+		  << " models)" << std::endl;
+
+	prepareTestData (numUsers, numEntriesPerUser, numTagsPerEntry);
+
+	if (cached == 1)
+	{
+		std::cout << "-> Loading data into cache" << std::endl;
+		(void)service::StorageEngine::instance ();
+	}
+
+	std::cout << "-> Executing benchmark:" << std::endl;
+
+	// decide what key to look for
+	int user_index = numUsers / 2;
+	int entry_index = numEntriesPerUser / 2;
+	std::string organization = "organization-" + std::to_string (entry_index);
+	std::string application = "/application-" + std::to_string (entry_index);
+	std::string scope = "/scope-" + std::to_string (entry_index);
+	std::string slug = "/slug-" + std::to_string (entry_index) + std::string ("-user-") + std::to_string (user_index);
+	std::string key = organization + application + scope + slug;
+
+	std::vector<model::Entry> entries;
+
+	// timer start
+	Timer timer;
+	timer.start ();
+
+	// the stuff to benchmark
+	entries.push_back (service::StorageEngine::instance ().getEntry (key));
+
+	// stop timer here
+	timer.stop ();
+
+	// print found entries
+	printEntryVector (entries, 3);
+
+	// print timer result
+	timer.printStatistic (3);
+
+	std::cout << std::endl;
+}
+
+void benchmarkLookupMultipleByOrganization (int numUsers, int numEntriesPerUser, int numTagsPerEntry, int cached)
+{
+	std::cout << "Benchmark: Lookup multiple entries by organization\n"
+		  << "          (" << numUsers << " users à " << numEntriesPerUser << " entries with " << numTagsPerEntry << " tags each)"
+		  << std::endl;
+	std::cout << "==============================================================" << std::endl;
+
+	std::cout << "-> Refreshing database (clear)" << std::endl;
+
+	clearDatabase ();
+
+	std::cout << "-> Creating test data (" << (numUsers + numUsers * numEntriesPerUser + numUsers * numEntriesPerUser * numTagsPerEntry)
+		  << " models)" << std::endl;
+
+	prepareTestData (numUsers, numEntriesPerUser, numTagsPerEntry);
+
+	if (cached == 1)
+	{
+		std::cout << "-> Loading data into cache" << std::endl;
+		(void)service::StorageEngine::instance ();
+	}
+
+	std::cout << "-> Executing benchmark:" << std::endl;
+
+	// decide what key to look for
+	int entry_index = numEntriesPerUser / 2;
+	std::string organization = "organization-" + std::to_string (entry_index);
+
+	// timer start
+	Timer timer;
+	timer.start ();
+
+	// the stuff to benchmark
+	std::vector<model::Entry> entries = service::StorageEngine::instance ().getAllEntries ();
```

Comment:
> Is there a way to avoid the copy constructor?

If you copy it from some internal vector from StorageEngine to the user: seems to be hardly avoidable.

> By design I tried to keep the functions modular

The design seems to be nice (I did not look into details though), the remarks are only about the benchmark.